File: packages/hashline/src/recovery.ts
/**
 * Recovers stale section tags by proving that every anchored line still maps
 * to one unchanged, contiguous region in the current file, then replaying the
 * edit against that live content.
 *
 * Recovery fails closed when the target changed or became ambiguous. The
 * patcher then returns a mismatch with fresh context instead of guessing.
 */
import { diffLineRuns } from "@oh-my-pi/pi-natives";
import { applyEdits } from "./apply";
import { RECOVERY_EXTERNAL_WARNING, RECOVERY_LINE_REMAP_WARNING, RECOVERY_SESSION_CHAIN_WARNING } from "./messages";
import type { SnapshotStore } from "./snapshots";
import type { Anchor, ApplyResult, Clipboard, Edit } from "./types";

export interface RecoveryArgs {
	path: string;
	currentText: string;
	fileHash: string;
	edits: readonly Edit[];
	/** Shared clipboard register for `cut`/`paste` edits, threaded into the replay apply. */
	clipboard?: Clipboard;
}

export interface RecoveryResult {
	/** Post-recovery text. */
	text: string;
	/** First changed line (1-indexed) relative to the live `currentText`, or `undefined`. */
	firstChangedLine: number | undefined;
	/** Warnings collected during recovery, including the user-facing recovery banner. */
	warnings: string[];
}

function collectAnchorLines(edits: readonly Edit[]): number[] {
	const lines: number[] = [];
	for (const edit of edits) {
		for (const anchor of getEditAnchors(edit)) lines.push(anchor.line);
	}
	return lines;
}

function getEditAnchors(edit: Edit): Anchor[] {
	if (edit.kind === "delete") return [edit.anchor];
	// Recovery only ever receives already-resolved edits (no `block`); this arm
	// exists for type-exhaustiveness over the full `Edit` union.
	if (edit.kind === "block") return [edit.anchor];
	if (edit.kind === "cut") {
		// Every captured line is an anchor: changed interior content is unsafe.
		const anchors: Anchor[] = [];
		for (let line = edit.range.start.line; line <= edit.range.end.line; line++) anchors.push({ line });
		return anchors;
	}
	if (edit.kind === "paste") {
		if (edit.at.kind === "span") {
			const anchors: Anchor[] = [];
			for (let line = edit.at.range.start.line; line <= edit.at.range.end.line; line++) anchors.push({ line });
			return anchors;
		}
		const cursor = edit.at.cursor;
		return cursor.kind === "before_anchor" || cursor.kind === "after_anchor" ? [cursor.anchor] : [];
	}
	return edit.cursor.kind === "before_anchor" || edit.cursor.kind === "after_anchor" ? [edit.cursor.anchor] : [];
}

function buildLineMap(previousText: string, currentText: string): Map<number, number> {
	const changes = diffLineRuns(previousText, currentText);
	const map = new Map<number, number>();
	let previousLine = 1;
	let currentLine = 1;

	for (const change of changes) {
		const count = change.count;
		if (change.added) {
			currentLine += count;
			continue;
		}
		if (change.removed) {
			previousLine += count;
			continue;
		}
		for (let offset = 0; offset < count; offset++) {
			map.set(previousLine + offset, currentLine + offset);
		}
		previousLine += count;
		currentLine += count;
	}

	return map;
}

/** Values appearing two or more times in `lines`, for O(1) duplicate checks. */
function collectDuplicatedValues(lines: readonly string[]): Set<string> {
	const seen = new Set<string>();
	const duplicated = new Set<string>();
	for (const value of lines) {
		if (seen.has(value)) duplicated.add(value);
		else seen.add(value);
	}
	return duplicated;
}

interface AnchorNeighbors {
	/** Nearest non-anchor line below the anchor's run, or `undefined` at the file edge. */
	before: number | undefined;
	/** Nearest non-anchor line above the anchor's run, or `undefined` at the file edge. */
	after: number | undefined;
}

/**
 * Nearest non-anchor context line on each side of every anchor, computed in
 * one sweep over the sorted anchor set. Anchors in one contiguous run share
 * both neighbors (the lines just outside the run), so this replaces the
 * per-anchor directional walk across anchored ranges — O(anchors²) on a
 * large block replacement — with one O(anchors log anchors) pass.
 */
function computeAnchorNeighbors(anchorLines: ReadonlySet<number>, lineCount: number): Map<number, AnchorNeighbors> {
	const sorted = [...anchorLines].sort((a, b) => a - b);
	const neighbors = new Map<number, AnchorNeighbors>();
	for (let i = 0; i < sorted.length;) {
		let j = i;
		while (j + 1 < sorted.length && sorted[j + 1] === sorted[j] + 1) j++;
		const start = sorted[i];
		const end = sorted[j];
		const before = start - 1 >= 1 && start - 1 <= lineCount ? start - 1 : undefined;
		const after = end + 1 <= lineCount ? end + 1 : undefined;
		for (let k = i; k <= j; k++) neighbors.set(sorted[k], { before, after });
		i = j + 1;
	}
	return neighbors;
}

function validateDuplicateAnchorContext(
	line: number,
	mapped: number,
	neighbors: AnchorNeighbors,
	lineMap: ReadonlyMap<number, number>,
): boolean {
	let checked = false;
	const { before, after } = neighbors;
	if (before !== undefined) {
		checked = true;
		if (lineMap.get(before) !== mapped - (line - before)) return false;
	}
	if (after !== undefined) {
		checked = true;
		if (lineMap.get(after) !== mapped + (after - line)) return false;
	}
	return checked;
}

function validateUniqueAnchorContext(
	line: number,
	mapped: number,
	neighbors: AnchorNeighbors,
	lineMap: ReadonlyMap<number, number>,
): boolean {
	const offset = mapped - line;
	const { before, after } = neighbors;
	if (after !== undefined && lineMap.get(after) === after + offset) return true;
	return before !== undefined && lineMap.get(before) === before + offset;
}

function validateRemappedAnchorContext(
	previousText: string,
	currentText: string,
	lineMap: ReadonlyMap<number, number>,
	edits: readonly Edit[],
): boolean {
	const previousLines = previousText.split("\n");
	const currentLines = currentText.split("\n");
	const anchorLines = new Set(collectAnchorLines(edits));
	// Precompute once per validation pass: which line values are duplicated,
	// and each anchor's nearest non-anchor context. The per-anchor forms —
	// indexOf/lastIndexOf full-file scans plus directional walks across
	// anchored ranges — are O(anchors×lines) + O(anchors²) and blow up on
	// large block replacements.
	const duplicatedPrevious = collectDuplicatedValues(previousLines);
	const duplicatedCurrent = collectDuplicatedValues(currentLines);
	const anchorNeighbors = computeAnchorNeighbors(anchorLines, previousLines.length);

	for (const [line, neighbors] of anchorNeighbors) {
		const mapped = lineMap.get(line);
		if (mapped === undefined) return false;
		if (!duplicatedPrevious.has(previousLines[line - 1]) && !duplicatedCurrent.has(currentLines[mapped - 1])) {
			if (!validateUniqueAnchorContext(line, mapped, neighbors, lineMap)) {
				return false;
			}
			continue;
		}
		if (!validateDuplicateAnchorContext(line, mapped, neighbors, lineMap)) {
			return false;
		}
	}

	return true;
}

interface RemappedEdits {
	edits: Edit[];
	offset: number;
}

function remapEditsToCurrent(previousText: string, currentText: string, edits: readonly Edit[]): RemappedEdits | null {
	const lineMap = buildLineMap(previousText, currentText);
	if (!validateRemappedAnchorContext(previousText, currentText, lineMap, edits)) return null;
	const offsets: number[] = [];

	const mapLine = (line: number): number | null => {
		const mapped = lineMap.get(line);
		if (mapped === undefined) return null;
		offsets.push(mapped - line);
		return mapped;
	};

	const mapAnchor = (anchor: Anchor): Anchor | null => {
		const line = mapLine(anchor.line);
		return line === null ? null : { line };
	};

	const remapped: Edit[] = [];
	for (const edit of edits) {
		if (edit.kind === "delete") {
			const anchor = mapAnchor(edit.anchor);
			if (anchor === null) return null;
			remapped.push({ ...edit, anchor });
			continue;
		}
		if (edit.kind === "block") {
			const anchor = mapAnchor(edit.anchor);
			if (anchor === null) return null;
			remapped.push({ ...edit, anchor });
			continue;
		}
		if (edit.kind === "cut") {
			// Map every captured line; an unmapped interior line means the
			// content drifted and cannot be moved safely. Uniform offsets keep
			// the mapped range contiguous.
			const start = mapLine(edit.range.start.line);
			if (start === null) return null;
			let end = start;
			for (let line = edit.range.start.line + 1; line <= edit.range.end.line; line++) {
				const mapped = mapLine(line);
				if (mapped === null) return null;
				end = mapped;
			}
			remapped.push({ ...edit, range: { start: { line: start }, end: { line: end } } });
			continue;
		}
		if (edit.kind === "paste") {
			let blockStart = edit.blockStart;
			if (blockStart !== undefined) {
				const mappedBlockStart = mapLine(blockStart);
				if (mappedBlockStart === null) return null;
				blockStart = mappedBlockStart;
			}
			if (edit.at.kind === "span") {
				const start = mapLine(edit.at.range.start.line);
				if (start === null) return null;
				let end = start;
				for (let line = edit.at.range.start.line + 1; line <= edit.at.range.end.line; line++) {
					const mapped = mapLine(line);
					if (mapped === null) return null;
					end = mapped;
				}
				remapped.push({
					...edit,
					at: { kind: "span", range: { start: { line: start }, end: { line: end } } },
					blockStart,
				});
				continue;
			}
			const cursor = edit.at.cursor;
			if (cursor.kind !== "before_anchor" && cursor.kind !== "after_anchor") {
				remapped.push(blockStart === edit.blockStart ? edit : { ...edit, blockStart });
				continue;
			}
			const anchor = mapAnchor(cursor.anchor);
			if (anchor === null) return null;
			remapped.push({ ...edit, at: { kind: "gap", cursor: { kind: cursor.kind, anchor } }, blockStart });
			continue;
		}
		if (edit.kind === "insert") {
			let blockStart = edit.blockStart;
			if (blockStart !== undefined) {
				const mappedBlockStart = mapLine(blockStart);
				if (mappedBlockStart === null) return null;
				blockStart = mappedBlockStart;
			}
			const cursor = edit.cursor;
			if (cursor.kind !== "before_anchor" && cursor.kind !== "after_anchor") {
				remapped.push(blockStart === edit.blockStart ? edit : { ...edit, blockStart });
				continue;
			}
			const anchor = mapAnchor(cursor.anchor);
			if (anchor === null) return null;
			remapped.push({ ...edit, cursor: { kind: cursor.kind, anchor }, blockStart });
		}
	}

	if (offsets.length === 0) return null;
	const firstOffset = offsets[0];
	if (!offsets.every(offset => offset === firstOffset)) return null;
	return { edits: remapped, offset: firstOffset };
}

function replayRemappedAnchorsOnCurrent(
	previousText: string,
	currentText: string,
	edits: readonly Edit[],
	recoveryWarning: string,
	clipboard: Clipboard | undefined,
	path: string,
): RecoveryResult | null {
	const remapped = remapEditsToCurrent(previousText, currentText, edits);
	if (remapped === null) return null;
	let applied: ApplyResult;
	try {
		applied = applyEdits(currentText, remapped.edits, {
			...(clipboard === undefined ? {} : { clipboard }),
			path,
		});
	} catch {
		return null;
	}
	if (applied.text === currentText) return null;
	return {
		text: applied.text,
		firstChangedLine: applied.firstChangedLine,
		warnings: [remapped.offset === 0 ? recoveryWarning : RECOVERY_LINE_REMAP_WARNING, ...(applied.warnings ?? [])],
	};
}
/**
 * Stateless recovery driver over a {@link SnapshotStore}. Construct once and
 * call {@link Recovery.tryRecover} per stale-tag incident.
 *
 * Recovery maps every stale anchor through unchanged lines from the tagged
 * snapshot to the live text, validates surrounding context, and replays the
 * edit directly on live content. All anchors must move by one consistent
 * offset. A changed, deleted, split, or ambiguous target is rejected so the
 * caller can surface a {@link MismatchError} with current context.
 */
export class Recovery {
	constructor(readonly store: SnapshotStore) {}
	/**
	 * Attempt recovery. Returns `null` when no path forward is found — the
	 * caller should then surface a {@link MismatchError}.
	 */
	tryRecover(args: RecoveryArgs): RecoveryResult | null {
		const { path, currentText, fileHash, edits, clipboard } = args;
		// When retained texts collide on the 16-bit tag, use the latest one.
		// Recovery still requires its anchors and context to map unambiguously.
		const snapshot = this.store.byHash(path, fileHash);
		if (!snapshot) return null;
		const recoveryWarning =
			this.store.head(path) === snapshot ? RECOVERY_EXTERNAL_WARNING : RECOVERY_SESSION_CHAIN_WARNING;
		return replayRemappedAnchorsOnCurrent(snapshot.text, currentText, edits, recoveryWarning, clipboard, path);
	}
}

File: crates/pi-edit/src/modes/hashline/recovery.rs
//! Stale-tag anchor recovery (`packages/hashline/src/recovery.ts`).

use std::{
	collections::{HashMap, HashSet},
	path::Path,
	sync::Arc,
};

use super::{
	apply::{ApplyOptions, EmptyPaste, apply_edits},
	messages::{
		RECOVERY_EXTERNAL_WARNING, RECOVERY_LINE_REMAP_WARNING, RECOVERY_SESSION_CHAIN_WARNING,
	},
	types::{Anchor, Cursor, Edit, ParsedRange, PasteTarget},
};
use crate::{
	error::EditError,
	store::{Clipboard, EditStore},
};

/// Inputs for stale-snapshot recovery.
pub struct RecoveryArgs<'a> {
	/// Canonical target path.
	pub path:         &'a Path,
	/// Current LF-normalized file text.
	pub current_text: &'a str,
	/// Requested stale snapshot tag.
	pub file_hash:    &'a str,
	/// Parsed edits anchored against the stale snapshot.
	pub edits:        &'a [Edit],
	/// Transactional clipboard, when the patch uses cut/paste.
	pub clipboard:    Option<&'a mut Clipboard>,
}

/// Successfully recovered application result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RecoveryResult {
	/// Recovered post-edit text.
	pub text:               String,
	/// First changed line in the live file.
	pub first_changed_line: Option<u32>,
	/// Recovery banner followed by apply-time warnings.
	pub warnings:           Vec<String>,
}

fn edit_anchors(edit: &Edit) -> Vec<Anchor> {
	match edit {
		Edit::Delete { anchor, .. } | Edit::Block { anchor, .. } => vec![*anchor],
		Edit::Cut { range, .. } => (range.start.line..=range.end.line)
			.map(|line| Anchor { line })
			.collect(),
		Edit::Paste { at: PasteTarget::Span { range }, .. } => (range.start.line..=range.end.line)
			.map(|line| Anchor { line })
			.collect(),
		Edit::Paste { at: PasteTarget::Gap { cursor }, .. } | Edit::Insert { cursor, .. } => {
			match cursor {
				Cursor::BeforeAnchor(anchor) | Cursor::AfterAnchor(anchor) => vec![*anchor],
				Cursor::Bof | Cursor::Eof => Vec::new(),
			}
		},
	}
}

fn build_line_map(previous: &str, current: &str) -> HashMap<u32, u32> {
	let mut map = HashMap::new();
	let mut previous_line = 1_u32;
	let mut current_line = 1_u32;
	for run in pi_diff::line_runs_str(previous, current) {
		if run.added {
			current_line += run.count;
		} else if run.removed {
			previous_line += run.count;
		} else {
			for offset in 0..run.count {
				map.insert(previous_line + offset, current_line + offset);
			}
			previous_line += run.count;
			current_line += run.count;
		}
	}
	map
}

fn duplicated_values(lines: &[&str]) -> HashSet<String> {
	let mut seen = HashSet::new();
	let mut duplicated = HashSet::new();
	for line in lines {
		if !seen.insert(*line) {
			duplicated.insert((*line).to_owned());
		}
	}
	duplicated
}

#[derive(Clone, Copy)]
struct Neighbors {
	before: Option<u32>,
	after:  Option<u32>,
}

fn anchor_neighbors(anchor_lines: &HashSet<u32>, line_count: usize) -> HashMap<u32, Neighbors> {
	let mut sorted: Vec<_> = anchor_lines.iter().copied().collect();
	sorted.sort_unstable();
	let mut neighbors = HashMap::new();
	let mut i = 0;
	while i < sorted.len() {
		let mut j = i;
		while j + 1 < sorted.len() && sorted[j + 1] == sorted[j] + 1 {
			j += 1;
		}
		let start = sorted[i];
		let end = sorted[j];
		let before = (start > 1).then_some(start - 1);
		let after = ((end as usize) < line_count).then_some(end + 1);
		for &line in &sorted[i..=j] {
			neighbors.insert(line, Neighbors { before, after });
		}
		i = j + 1;
	}
	neighbors
}

fn validate_context(
	previous: &str,
	current: &str,
	line_map: &HashMap<u32, u32>,
	edits: &[Edit],
) -> bool {
	let previous_lines: Vec<_> = previous.split('\n').collect();
	let current_lines: Vec<_> = current.split('\n').collect();
	let anchors: HashSet<_> = edits
		.iter()
		.flat_map(edit_anchors)
		.map(|anchor| anchor.line)
		.collect();
	let previous_duplicates = duplicated_values(&previous_lines);
	let current_duplicates = duplicated_values(&current_lines);
	for (line, neighbors) in anchor_neighbors(&anchors, previous_lines.len()) {
		let Some(&mapped) = line_map.get(&line) else {
			return false;
		};
		let previous_value = previous_lines
			.get((line - 1) as usize)
			.copied()
			.unwrap_or("");
		let current_value = current_lines
			.get((mapped - 1) as usize)
			.copied()
			.unwrap_or("");
		if !previous_duplicates.contains(previous_value)
			&& !current_duplicates.contains(current_value)
		{
			let offset = i64::from(mapped) - i64::from(line);
			let after_matches = neighbors.after.is_some_and(|after| {
				line_map
					.get(&after)
					.is_some_and(|&value| i64::from(value) == i64::from(after) + offset)
			});
			let before_matches = neighbors.before.is_some_and(|before| {
				line_map
					.get(&before)
					.is_some_and(|&value| i64::from(value) == i64::from(before) + offset)
			});
			if !after_matches && !before_matches {
				return false;
			}
		} else {
			let mut checked = false;
			if let Some(before) = neighbors.before {
				checked = true;
				if line_map.get(&before).copied() != mapped.checked_sub(line - before) {
					return false;
				}
			}
			if let Some(after) = neighbors.after {
				checked = true;
				if line_map.get(&after).copied() != Some(mapped + (after - line)) {
					return false;
				}
			}
			if !checked {
				return false;
			}
		}
	}
	true
}

fn map_line(line_map: &HashMap<u32, u32>, line: u32, offsets: &mut Vec<i64>) -> Option<u32> {
	let mapped = *line_map.get(&line)?;
	offsets.push(i64::from(mapped) - i64::from(line));
	Some(mapped)
}

fn remap_edits(previous: &str, current: &str, edits: &[Edit]) -> Option<(Vec<Edit>, i64)> {
	let line_map = build_line_map(previous, current);
	if !validate_context(previous, current, &line_map, edits) {
		return None;
	}
	let mut offsets = Vec::new();
	let mut remapped = Vec::with_capacity(edits.len());
	for edit in edits {
		let mapped = match edit {
			Edit::Delete { anchor, line_num, index, old_assertion } => Edit::Delete {
				anchor:        Anchor { line: map_line(&line_map, anchor.line, &mut offsets)? },
				line_num:      *line_num,
				index:         *index,
				old_assertion: old_assertion.clone(),
			},
			Edit::Block { anchor, payloads, mode, register, line_num, index } => Edit::Block {
				anchor:   Anchor { line: map_line(&line_map, anchor.line, &mut offsets)? },
				payloads: payloads.clone(),
				mode:     *mode,
				register: register.clone(),
				line_num: *line_num,
				index:    *index,
			},
			Edit::Cut { range, register, line_num, index } => {
				let start = map_line(&line_map, range.start.line, &mut offsets)?;
				let mut end = start;
				for line in range.start.line + 1..=range.end.line {
					end = map_line(&line_map, line, &mut offsets)?;
				}
				Edit::Cut {
					range:    ParsedRange { start: Anchor { line: start }, end: Anchor { line: end } },
					register: register.clone(),
					line_num: *line_num,
					index:    *index,
				}
			},
			Edit::Paste { at, register, line_num, index, block_start } => {
				let block_start = match block_start {
					Some(line) => Some(map_line(&line_map, *line, &mut offsets)?),
					None => None,
				};
				let at = match at {
					PasteTarget::Span { range } => {
						let start = map_line(&line_map, range.start.line, &mut offsets)?;
						let mut end = start;
						for line in range.start.line + 1..=range.end.line {
							end = map_line(&line_map, line, &mut offsets)?;
						}
						PasteTarget::Span {
							range: ParsedRange {
								start: Anchor { line: start },
								end:   Anchor { line: end },
							},
						}
					},
					PasteTarget::Gap { cursor: Cursor::BeforeAnchor(anchor) } => PasteTarget::Gap {
						cursor: Cursor::BeforeAnchor(Anchor {
							line: map_line(&line_map, anchor.line, &mut offsets)?,
						}),
					},
					PasteTarget::Gap { cursor: Cursor::AfterAnchor(anchor) } => PasteTarget::Gap {
						cursor: Cursor::AfterAnchor(Anchor {
							line: map_line(&line_map, anchor.line, &mut offsets)?,
						}),
					},
					PasteTarget::Gap { cursor } => PasteTarget::Gap { cursor: *cursor },
				};
				Edit::Paste {
					at,
					register: register.clone(),
					line_num: *line_num,
					index: *index,
					block_start,
				}
			},
			Edit::Insert { cursor, text, line_num, index, replacement, block_start } => {
				let block_start = match block_start {
					Some(line) => Some(map_line(&line_map, *line, &mut offsets)?),
					None => None,
				};
				let cursor = match cursor {
					Cursor::BeforeAnchor(anchor) => Cursor::BeforeAnchor(Anchor {
						line: map_line(&line_map, anchor.line, &mut offsets)?,
					}),
					Cursor::AfterAnchor(anchor) => Cursor::AfterAnchor(Anchor {
						line: map_line(&line_map, anchor.line, &mut offsets)?,
					}),
					other => *other,
				};
				Edit::Insert {
					cursor,
					text: text.clone(),
					line_num: *line_num,
					index: *index,
					replacement: *replacement,
					block_start,
				}
			},
		};
		remapped.push(mapped);
	}
	let (&first, rest) = offsets.split_first()?;
	if rest.iter().any(|offset| *offset != first) {
		return None;
	}
	Some((remapped, first))
}

/// Attempt to rebase stale anchors from a retained snapshot onto current text.
pub fn try_recover(
	store: &EditStore,
	mut args: RecoveryArgs<'_>,
) -> Result<Option<RecoveryResult>, EditError> {
	let Some(snapshot) = store.by_hash(args.path, args.file_hash) else {
		return Ok(None);
	};
	let warning = if store
		.head(args.path)
		.is_some_and(|head| Arc::ptr_eq(&head.text, &snapshot.text))
	{
		RECOVERY_EXTERNAL_WARNING
	} else {
		RECOVERY_SESSION_CHAIN_WARNING
	};
	let Some((edits, offset)) = remap_edits(&snapshot.text, args.current_text, args.edits) else {
		return Ok(None);
	};
	let Ok(applied) = apply_edits(args.current_text, &edits, ApplyOptions {
		clipboard:      args.clipboard.take(),
		path:           args.path.to_str(),
		on_empty_paste: EmptyPaste::Throw,
	}) else {
		return Ok(None);
	};
	if applied.text == args.current_text {
		return Ok(None);
	}
	let mut warnings = vec![if offset == 0 {
		warning.to_owned()
	} else {
		RECOVERY_LINE_REMAP_WARNING.to_owned()
	}];
	warnings.extend(applied.warnings);
	Ok(Some(RecoveryResult {
		text: applied.text,
		first_changed_line: applied.first_changed_line,
		warnings,
	}))
}
